modules/packages/Sort.chpl

```diff
@@ -18,53 +18,321 @@
+  record Comparator { }
+
+  // compare method defines how 2 elements are compared
+  proc Comparator.compare(a, b) {
+    return abs(a) - abs(b);
+  }
+
+  var absReverseComparator: ReverseComparator(Comparator);
+
+  QuickSort(Array, comparator=absReverseComparator);
+
+  // This will output: -4, 3, 2, -1
+  writeln(Array);
 
  */
 module Sort {
 
-private inline proc chpl_sort_cmp(a, b, param reverse=false, param eq=false) {
-  if eq {
-    if reverse then return a >= b;
-    else return a <= b;
+
+/* Module-defined comparators */
+
+/*
+  Instance of :record:`DefaultComparator` used as default ``comparator=``
+  argument when no comparator is passed to a sort function
+*/
+const defaultcomparator: DefaultComparator;
+
+
+
+/*
+   Instance of :record:`ReverseComparator`. Pass this as the ``comparator=``
+   argument of a sort function to reverse the sort order.
+ */
+const reversecomparator: ReverseComparator(DefaultComparator);
+
+
+/* Private methods */
+
+/* Base compare method of all sort functions */
+private inline proc chpl_compare(a, b, comparator:?rec=defaultcomparator) {
+  use Reflection;
+
+  // TODO -- In cases where values are larger than keys, it may be faster to
+  //         key data once and sort the keyed data, mirroring swaps in data.
+  // Compare results of comparator.key(a) if is defined by user
+  if canResolveMethod(comparator, "key", a) && canResolveMethod(comparator, "key", b) {
+    return chpl_compare(comparator.key(a), comparator.key(b));
+  // Use comparator.compare(a, b) if is defined by user
+  } else if canResolveMethod(comparator, "compare", a, b) {
+    return comparator.compare(a ,b);
   } else {
-    if reverse then return a > b;
-    else return a < b;
+    compilerError("The comparator record requires a 'key(a)' or 'compare(a, b)' method");
   }
 }
 
 
 /*
+    Check if a comparator was passed and confirm that it will work, otherwise
+    throw a compile-time error.
+
+    :arg a: Sample data passed to confirm that comparator methods can resolve
+    :arg comparator: Record that redefines the comparison mechanism with one of
+      the methods: ``comparator.key(a)`` or ``comparator.compare(a,b)``
+
+ */
+private proc chpl_check_comparator(comparator, type eltType) {
+  use Reflection;
+
+  // Dummy data for checking method resolution
+  // This may need updating when constructors support non-default args
+  const data: eltType;
+
+  if comparator.type == DefaultComparator {}
+  // Check for valid comaparator methods
+  else if canResolveMethod(comparator, "compare", data, data) {
+    // Check return type of compare
+    type comparetype = comparator.compare(data, data).type;
+    if !(isNumericType(comparetype)) {
+      compilerError("The compare method must return a numeric type");
+    }
+  }
+  else if canResolveMethod(comparator, "key", data) {
+    // Check return type of key
+    type keytype = comparator.key(data).type;
+    if !(isNumericType(keytype) || isStringType(keytype)) {
```

Comment:
The comments above indicate 'key must support numeric operations'. Should it say 'key must support numeric operations or be a string' ?